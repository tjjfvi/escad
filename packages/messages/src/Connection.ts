
export interface Connection<T> {
  send: (value: T) => void,
  onMsg: (cb: (value: T) => void) => void,
  offMsg: (cb: (value: T) => void) => void,
}
