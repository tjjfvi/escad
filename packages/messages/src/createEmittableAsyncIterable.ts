
export const createEmittableAsyncIterable = <T>() => {
  let listeners = new Set<(value: T) => void>();

  return [
    (value: T) => {
      const oldListeners = listeners;
      listeners = new Set();
      oldListeners.forEach(fn => fn(value));
    },
    async function*(): AsyncIterable<T>{
      while(true)
        yield await new Promise(res => listeners.add(res))
    }
  ] as const;
}
