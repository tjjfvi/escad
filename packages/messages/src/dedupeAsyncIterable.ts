
export function dedupeAsyncIterable<T>(iterableFn: () => AsyncIterable<T>){
  let iterator: AsyncIterator<T> | undefined
  let curPromise: Promise<IteratorResult<T>> | undefined
  const dedupedIterable = {
    [Symbol.asyncIterator]: () => ({
      next: () => {
        iterator ??= iterableFn()[Symbol.asyncIterator]()
        curPromise ??= iterator.next().finally(() => curPromise = undefined)
        return curPromise
      },
    }),
  }
  const fn = async function*(){
    for await (const value of dedupedIterable)
      yield value
  }
  fn[Symbol.asyncIterator] = fn
  return fn
}
