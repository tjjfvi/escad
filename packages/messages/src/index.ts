
// @create-index {"mode":"*"}

export * from './Connection';
export * from './Messenger';
export * from './brandConnection';
export * from './childProcessConnection';
export * from './createMessenger';
export * from './filterConnection';
export * from './mapConnection';
export * from './noopConnection';
export * from './workerConnection';
export * from './wsConnection';

