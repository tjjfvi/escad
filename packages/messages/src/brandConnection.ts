
import { Connection } from "./Connection";
import { filterConnection } from "./filterConnection";
import { mapConnection } from "./mapConnection";

export const brandConnection =
  (connection: Connection<unknown>, brand: string): Connection<unknown> =>
    mapConnection(
      filterConnection(
        connection,
        (v): v is { brand: string, value: unknown } =>
          typeof v === "object" && !!v && "brand" in v && v["brand" as never] === brand && "value" in v
      ),
      value => ({ brand, value }),
      ({ value }) => value,
    )
