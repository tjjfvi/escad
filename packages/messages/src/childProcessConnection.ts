
import type { ChildProcess } from "child_process"
import { Connection } from "./Connection"

/* istanbul ignore next: covered by types, difficult to test */
export const childProcessConnection = (childProcess: ChildProcess): Connection<unknown> => ({
  send: value => childProcess.send(value as never),
  onMsg: cb => {
    childProcess.on("message", cb)
    return () => childProcess.off("message", cb)
  },
  destroy: () => childProcess.kill(),
})
