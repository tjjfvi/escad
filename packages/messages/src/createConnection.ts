import { Connection } from "./Connection";

export const createConnection = <T>(send: (value: T) => void): Connection<T> & { emit: (value: T) => void } => {
  const listeners = new Set<(value: T) => void>();
  return {
    send,
    onMsg: cb => listeners.add(cb),
    offMsg: cb => listeners.delete(cb),
    emit: value => listeners.forEach(cb => cb(value)),
  }
}
