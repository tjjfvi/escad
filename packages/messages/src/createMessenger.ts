
import { Connection } from "./Connection";
import { Messenger, MessengerImpl, MessengerShape } from "./Messenger";

export const createMessenger = (
  <F extends MessengerShape, T extends MessengerShape>(
    impl: MessengerImpl<F, T>,
    socket: Connection<unknown>,
  ): Messenger<F, T> => {
    let idN = 0;
    const resolveMap = Object.create(null)
    const rejectMap = Object.create(null)
    const iteratorMap = Object.create(null)
    const other = new Proxy(Object.create(null), {
      get: (target, prop) => {
        if(prop in target)
          return target[prop];
        if(typeof prop === "symbol")
          return;
        return target[prop] = (...args: any[]) => {
          const id = ++idN
          socket.send(["call", id, prop, ...args]);
          return Object.assign(
            recvPromise(id),
            {
              [Symbol.asyncIterator]: () => ({
                next: () => {
                  const id2 = ++idN
                  socket.send(["next", id, id2]);
                  return recvPromise(id2);
                },
              })
            }
          );
        };
      }
    });
    (impl as any).req = other;
    socket.onMsg(msg => {
      if(!msg || !(msg instanceof Array))
        return;
      const [kind, id, key, ...args] = msg;
      if(typeof id !== "number" || typeof key !== "number" && typeof key !== "string" && key !== null)
        return;
      switch(kind) {
        case "call": {
          const result = impl[key]?.(...args) ?? Promise.reject(new Error(`No method with the key "${key}"`));
          if("then" in result)
            sendPromise(id, result);
          else if(Symbol.asyncIterator in result)
            iteratorMap[id] = result[Symbol.asyncIterator]();
          return;
        }
        case "next": {
          sendPromise(+key, iteratorMap[id].next());
          return;
        }
        case "resolve": {
          if(id in resolveMap)
            resolveMap[id](args[0]);
          delete resolveMap[id];
          return;
        }
        case "reject": {
          if(id in rejectMap)
            rejectMap[id](args[0]);
          delete rejectMap[id];
          return;
        }
      }
    })
    return impl as any;

    function recvPromise(id: number){
      return new Promise((resolve, reject) => {
        resolveMap[id] = resolve;
        rejectMap[id] = reject
      });
    }

    function sendPromise(id: number, promise: Promise<unknown>){
      promise.then(
        value =>
          socket.send(["resolve", id, null, value]),
        value =>
          socket.send(["reject", id, null, value]),
      )
    }
  }
)
