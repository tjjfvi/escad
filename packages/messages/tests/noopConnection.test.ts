
import { noopConnection } from "../src"

test("", () => {
  const [a, b] = noopConnection<number>()
  const fn0 = jest.fn()
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  const fn3 = jest.fn()
  a.send(-1)
  a.onMsg(fn0)
  a.send(-1)
  b.send(0)
  b.onMsg(fn1)
  a.send(1)
  b.onMsg(fn2)
  a.send(2)
  a.onMsg(fn3)
  b.onMsg(fn3)
  a.send(3)
  b.send(4)
  a.offMsg(fn0)
  a.offMsg(fn1)
  b.send(5)
  b.offMsg(fn1)
  a.send(6)
  expect([fn0, fn1, fn2, fn3].map(f => f.mock.calls)).toMatchSnapshot()
})
