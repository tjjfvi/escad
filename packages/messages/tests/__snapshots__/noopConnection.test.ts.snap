// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  Array [
    Array [
      0,
    ],
    Array [
      4,
    ],
  ],
  Array [
    Array [
      1,
    ],
    Array [
      2,
    ],
    Array [
      3,
    ],
  ],
  Array [
    Array [
      2,
    ],
    Array [
      3,
    ],
    Array [
      6,
    ],
  ],
  Array [
    Array [
      3,
    ],
    Array [
      4,
    ],
    Array [
      5,
    ],
    Array [
      6,
    ],
  ],
]
`;
